File: piggybank/core/src/ports/payments.rs
//! Persistence + read ports for the [`PaymentOrder`] aggregate.
//!
//! Every method is internally atomic and the method IS the transaction boundary — no caller
//! ever holds one across the port. Two of them carry the weight:
//!
//! - [`PaymentRepository::open`] writes the order, its materialized approval requirement
//!   (a consent seat, or the link to the consilium that will decide it) and the drained
//!   events in ONE transaction, so an order can never exist with no record of what would
//!   authorize it;
//! - [`PaymentRepository::submit`] counts the attempt, compares the code, records the
//!   decision and transitions the order **all under one `SELECT … FOR UPDATE` on the
//!   `payments` row** — not on the consent row, because the order is what the transition
//!   belongs to and locking the seat instead would let the approval and the order move in
//!   two transactions a retry could interleave.
//!
//! **TigerBeetle is never touched here.** A payment's money moves in the relay, from the
//! `Reserved`/`Settled` events these methods drain to the outbox — the one ACID point stays
//! Postgres, exactly as it does for every other money aggregate.

use async_trait::async_trait;
use domain::{
	balance::Party,
	consilium::ConsiliumId,
	error::DomainError,
	payments::{PaymentEffect, PaymentId, PaymentOrder, PaymentState},
	users::UserId,
};

/// The digest length every stored secret is reduced to, named so the schema's
/// `octet_length(...) = 32` checks and the Rust side cannot drift apart.
pub const DIGEST_BYTES: usize = 32;

/// How many wrong codes a consent seat may submit before its token burns permanently — and,
/// with it, the payment fails closed. The consilium's ceiling, deliberately: `docs/CONSILIUM.md`
/// § "One specification for both planes" is one specification, not shared code.
pub const MAX_CODE_ATTEMPTS: i32 = 5;

/// What must be written beside a brand-new order to materialize its one approval requirement.
///
/// A sum type rather than two optional fields: an order needs exactly one of these, decided by
/// [`domain::payments::PaymentTerms::requirement`], and a shape that could carry both or
/// neither would let an unauthorizable order commit.
pub enum ApprovalSeat {
	/// The source is fund-owned: the owner quorum that decides this order. The consilium
	/// itself is opened through [`ConsiliumRepository`](super::ConsiliumRepository); this is
	/// only the link, recorded so the order names its authorization from the moment it exists.
	Consilium(ConsiliumId),
	/// The source is one investor's claim: that investor's own emailed consent.
	Consent(ConsentCredential),
}

/// The minted consent seat. The plaintext token and code travel no further than the mail row
/// [`PaymentRepository::open`] writes from them — in the SAME transaction as the seat, so a
/// concierge outage can never leave an order nobody was asked about, and a mail can never
/// exist for an order that failed to commit. Only their digests are stored on the seat, so a
/// dump of `payment_consent` yields nothing that can consent to anything.
pub struct ConsentCredential {
	pub subject: UserId,
	/// The opaque single-use token that goes in the emailed link.
	pub token: String,
	/// The secret code the subject types on the consent page.
	pub code: String,
	pub token_hash: [u8; DIGEST_BYTES],
	pub code_hash: [u8; DIGEST_BYTES],
	/// The subject's folded revoke floor at open — `GREATEST(concierge_token_version,
	/// token_version)`, the same value [`IssuanceTarget::token_version`](super::IssuanceTarget)
	/// carries, so a revoke on either plane counts. Re-checked at consent and again at
	/// execution, which is what makes `RevokeTokens` void a consent that is already in flight.
	pub token_version_at_open: u64,
	/// SHA-256 over the subject's mirrored `users.email` at open, byte for byte as stored
	/// (already normalized by `Email::parse`). Re-checked at consent and at execution, so
	/// changing the mailbox at the identity provider cannot redirect a live token.
	pub email_hash_at_open: [u8; DIGEST_BYTES],
}

/// The consent seat as a surface renders it.
#[derive(Debug)]
pub struct ConsentView {
	pub subject: UserId,
	pub email: String,
	pub decision: ConsentDecision,
	/// Unix seconds; 0 while pending.
	pub decided_at: i64,
	pub notified: bool,
	pub attempts_remaining: u32,
	/// Why the seat can no longer be answered or executed — one of the pins recorded at open
	/// has moved — or `None` while it still holds. The execution path for an L1 order must
	/// read this BEFORE creating the withdrawal: `record_execution` refuses a moved pin too,
	/// but by then the withdrawal would already exist.
	pub invalidated: Option<String>,
}

/// What the emailed investor answered. The consilium's `VoteDecision` is the same three
/// values and is deliberately NOT reused: these are two planes that happen to agree today,
/// and folding them would make a future divergence a refactor of both.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ConsentDecision {
	Pending,
	Approve,
	Reject,
}

impl ConsentDecision {
	pub fn as_str(self) -> &'static str {
		match self {
			Self::Pending => "pending",
			Self::Approve => "approve",
			Self::Reject => "reject",
		}
	}

	pub fn parse(raw: &str) -> Result<Self, DomainError> {
		match raw {
			"pending" => Ok(Self::Pending),
			"approve" => Ok(Self::Approve),
			"reject" => Ok(Self::Reject),
			other => Err(DomainError::Validation(format!("unknown consent decision: {other}"))),
		}
	}
}

/// What a human recognises the receiving end BY, beside the canonical label. The label
/// (`PaymentTerms::destination_label`) is what the digest binds and every surface shares; an
/// investor's masked mailbox or a product's title is the detail that lets an approver tell
/// "investor 8f3e…" from the person they meant. Never digested, and masked by the surface.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum EndDetail {
	/// The receiving investor's mirrored address, UNMASKED — every surface masks it.
	Mailbox(String),
	/// The receiving product's title from the allocation registry.
	ProductTitle(String),
}

/// An order and the identity slice a surface needs beside it.
#[derive(Debug)]
pub struct PaymentView {
	pub order: PaymentOrder,
	pub initiator_email: String,
	/// Present exactly when the requirement is the owner quorum.
	pub consilium_id: Option<ConsiliumId>,
	/// Present exactly when the requirement is the subject's consent.
	pub consent: Option<ConsentView>,
	/// The receiving end's recognisable detail, when it has one.
	pub destination_detail: Option<EndDetail>,
}

/// What the emailed investor is shown. Deliberately narrower than [`PaymentView`]: the terms
/// they are consenting to and nothing about the operator who proposed it beyond an address.
#[derive(Debug)]
pub struct ConsentInvitation {
	pub payment_id: PaymentId,
	pub state: PaymentState,
	pub order: PaymentOrder,
	pub payload_hash: String,
	pub initiator_email: String,
	pub subject_email: String,
	pub expires_at: i64,
	pub decision: ConsentDecision,
	pub attempts_remaining: u32,
	pub destination_detail: Option<EndDetail>,
}

/// Audit facts the edge supplies with a consent. Recorded, never trusted for authorization.
pub struct ConsentAudit {
	pub client_ip: String,
	pub user_agent: String,
}

/// The result of a consent that was actually accepted.
#[derive(Debug)]
pub struct ConsentOutcome {
	pub payment: PaymentView,
	/// True when this answer carried the order out of `pending`.
	pub decided: bool,
	/// True when that answer was approval — the signal to attempt execution.
	pub approved: bool,
}

/// How an execution attempt ended. Written by [`PaymentRepository::record_execution`].
pub enum ExecutionOutcome {
	Executed(PaymentEffect),
	Failed(String),
}

/// Where an L2/L3 order's reservation stands in the relay — the Read-First the settlement
/// takes before it is recorded, because `Approved` commits the `Reserved` event and the
/// relay applies it afterwards, in its own time.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ReservationStatus {
	/// The relay applied the reserve: its deterministic transfer id is in `saga_steps`.
	Applied,
	/// Still in the outbox, or not yet reached. Try again later.
	Pending,
	/// The ledger refused the reserve and the relay parked the event. The order can never
	/// settle; an operator sees the row in the parked outbox.
	Parked,
}

/// What an admin screen filters the payments table by. Every field is additive and `None`
/// means "no restriction", so the empty filter is the whole history.
#[derive(Default)]
pub struct PaymentFilter {
	pub state: Option<PaymentState>,
	/// Either end. Most payments have no user on either side (`Piggybank → Revenue`), which
	/// is exactly why the admin screen cannot be served by the per-user operation feed.
	pub party: Option<Party>,
	/// Only orders whose source is fund-owned (the governance surface), when true.
	pub fund_owned_source: Option<bool>,
}

/// The refusal for a second open order against one fund-owned source claim — the partial
/// unique index's answer, and the pre-check's, in the same words: which of the two spoke
/// is an accident of timing the caller has no business telling apart.
pub fn already_open() -> DomainError {
	DomainError::Conflict("a payment is already open against this claim — close it before opening another".into())
}

/// The single response every unusable consent token produces, whatever made it unusable —
/// unknown, expired, spent, burned, or attached to an order that has closed. A caller cannot
/// tell which they hit, so the surface cannot be used to probe for live tokens.
pub fn consent_not_found() -> DomainError {
	DomainError::NotFound {
		entity: "consent",
		id: String::new(),
	}
}

#[async_trait]
pub trait PaymentRepository: Send + Sync {
	/// Persist a brand-new order together with its materialized approval requirement, in one
	/// transaction. Takes the source claim's write lock ([`lock_claim`](crate::infrastructure::outbox::lock_claim))
	/// first: a payment spends the same claims withdrawals and subscriptions do, and a
	/// spender that skips the lock silently stops the serialization covering the others.
	///
	/// Refuses with [`DomainError::Conflict`] when another order is already open against the
	/// same fund-owned source (the partial unique index is what actually enforces it).
	///
	/// For a consent seat the invitation mail is queued in this same transaction, addressed
	/// to the subject's identity-plane id; `consent_url_base` is what the emailed link is
	/// built on. Refused when the subject has no mirrored concierge id — there would be no
	/// safe address to send to, and an order nobody can be asked about must not exist.
	async fn open(&self, order: &mut PaymentOrder, seat: ApprovalSeat, consent_url_base: &str) -> Result<(), DomainError>;

	/// Load one order in full (no lock; for queries).
	async fn find(&self, id: PaymentId) -> Result<Option<PaymentView>, DomainError>;

	/// Whether an order is already open (`pending` or `approved`) against `source`. A read
	/// the fund-owned open path takes BEFORE seating its consilium: the unique index would
	/// refuse the order anyway, but only after a quorum had been opened and every owner
	/// mailed — and then withdrawn and every owner mailed again. Advisory, not the guard:
	/// the index is the guard.
	async fn has_open_against(&self, source: &Party) -> Result<bool, DomainError>;

	/// Record that the owner quorum carried this order — the consilium branch's counterpart
	/// to [`Self::submit`]. Applies [`PaymentOrder::approve`] under the order's row lock, so
	/// the reservation event and the state change commit together. Idempotent.
	///
	/// `consilium` must be the quorum this order was opened under (the `payment_approval`
	/// link), checked under the same lock: a consilium over one order must not be able to
	/// approve another, whatever its terms happen to say.
	async fn record_approval(&self, id: PaymentId, consilium: ConsiliumId, at: i64) -> Result<PaymentView, DomainError>;

	/// Record that the owner quorum refused, or that the order was withdrawn by its initiator.
	/// Under the row lock; idempotent on an already-closed order.
	async fn record_rejection(&self, id: PaymentId, at: i64) -> Result<PaymentView, DomainError>;

	/// Withdraw an order the caller opened, under the row lock. Refuses a caller who is not
	/// the initiator; idempotent on an already-cancelled one.
	async fn cancel(&self, id: PaymentId, by: UserId, at: i64) -> Result<PaymentView, DomainError>;

	/// Resolve an emailed consent token to its invitation. **Strictly side-effect free** —
	/// mail scanners issue automatic requests for every URL in a message, so this must not
	/// count an attempt, spend a token, or record anything.
	async fn invitation(&self, token_hash: &[u8; DIGEST_BYTES], at: i64) -> Result<ConsentInvitation, DomainError>;

	/// Answer a consent: count the attempt, compare the code in constant time, record the
	/// decision and transition the order — one transaction, one lock, taken on the `payments`
	/// row. A repeat of the same answer is an idempotent no-op; a different one is refused.
	///
	/// Burning the token (five wrong codes) **fails the payment closed**. With one seat there
	/// is no second party to escalate to, so the burn is a refusal, not a detector.
	///
	/// A seat whose pins have moved since open (the subject's sessions revoked, or their
	/// mailbox changed) is refused with [`DomainError::Conflict`] before the code is compared,
	/// and the order is rejected with it — fail-closed, and terminal for the same one-seat
	/// reason.
	async fn submit(&self, token_hash: &[u8; DIGEST_BYTES], code: &str, decision: ConsentDecision, audit: &ConsentAudit, at: i64) -> Result<ConsentOutcome, DomainError>;

	/// Expire every pending order past its deadline. Returns how many closed.
	async fn expire_due(&self, at: i64) -> Result<usize, DomainError>;

	/// Orders that are approved but have no effect yet — what an execution retry picks up
	/// after a crash between the approval and the money. An `execution_failed` one is
	/// deliberately NOT here: nothing retries silently.
	async fn awaiting_execution(&self) -> Result<Vec<PaymentId>, DomainError>;

	/// Whether the relay has applied this order's reservation (`reserve_tid` is its
	/// deterministic transfer id, `uuid_v5(payment_id, "payment:reserve")`). Read from the
	/// relay's own record, `saga_steps` and the parked outbox — not from the ledger, which
	/// cannot say "refused" — so the settlement is never recorded over a reserve that
	/// never landed.
	async fn reservation_status(&self, id: PaymentId, reserve_tid: u128) -> Result<ReservationStatus, DomainError>;

	/// Record how the execution attempt ended, under the row lock. Writing the effect is
	/// idempotent for the same effect and a conflict for a different one — which is what lets
	/// the caller re-read by the deterministic id and believe the row rather than the error.
	///
	/// An `Executed` outcome over a consent seat whose pins have moved since open is NOT
	/// recorded: the order is moved to `execution_failed` (releasing an L2/L3 reservation)
	/// and the call returns [`DomainError::Conflict`] naming why. See
	/// [`ConsentView::invalidated`] for the L1 path, which must check before the withdrawal
	/// exists — and for the window between that check and this call, a withdrawal the
	/// refused effect names is CANCELLED in this same transaction while it is still queued,
	/// so a revocation that lands after the check still stops the money. A withdrawal that
	/// has already been dispatched cannot be voided (the broadcast may have landed), so the
	/// effect is recorded as it is and the pin movement is logged rather than lied about.
	async fn record_execution(&self, id: PaymentId, outcome: ExecutionOutcome, at: i64) -> Result<PaymentView, DomainError>;
}

/// The admin payments screen's read model.
///
/// A **query-side** port, separate from [`PaymentRepository`] for the reason
/// [`OperationFeed`](super::OperationFeed) is separate from the aggregates it reads: it owns
/// nothing and writes nothing. It exists as its own trait rather than another method on the
/// repository so the screen's permission and paging can be granted without handing out the
/// write surface with them.
#[async_trait]
pub trait PaymentFeed: Send + Sync {
	/// The payment history matching `filter`, newest first, capped at `limit`. Nothing is
	/// ever deleted, so rejected, expired and failed orders stay in it — the audit record is
	/// half the point of the screen.
	async fn list(&self, filter: &PaymentFilter, limit: i64) -> Result<Vec<PaymentView>, DomainError>;
}

File: cabinet/frontend/shared/contracts/payments.ts
// The payments surface: a payment ORDER between two named ends of the platform, and the
// emailed consent an investor gives when the money is their own.
//
// Hand-written like `./governance`: these are the BFF's DTOs, shaped for this cabinet, not
// the proto messages `./gen` projects (`PaymentEnd.detail`, the nested `consent`, the
// lowercase states and the string timestamps are all the BFF's doing — docs/CONSILIUM.md
// § Payments). When the surface is ever served straight off the proto, this file is what
// `./gen` replaces.
//
// Two facts the money plane derives are carried here as READ-ONLY fields, never inputs:
//
//   · `tier` is a function of the destination (`external` for an address, `service` for a
//     product's pooled claim, `internal` for everything else). The open request carries no
//     tier; a caller-supplied one would only ever be interesting when it disagreed.
//   · `requirement` is a function of the SOURCE and nothing else: fund-owned money asks the
//     owner consilium, an investor's own claim asks that investor. There is no cell in
//     which one admin moves fund money alone, and no field here that could ask for one.
//
// The form previews both before the confirm click (`views/admin/payments/lib/terms.ts`)
// so the operator is not surprised by who gets emailed — but the plane's answer is the one
// that counts, and the list renders the plane's, not the preview.
//
// Lifecycle states travel as plain strings, for the reason `./governance` gives: the state
// machine lives in the money plane, and a closed union here would turn a backend addition
// into a client type error. Views label the states they know and fall back to the wire
// word.

import type { Decimal, Timestamp } from "./governance";

// ── Naming an end ──────────────────────────────────────────────────────────────

/** An internal claim as the caller names it when OPENING an order. */
export type PartyKind = "piggybank" | "revenue" | "service" | "user";

/**
 * One internal end of an order. `id` is empty for the two singleton claims (`piggybank`,
 * `revenue`), the product id for `service`, and the banking user id for `user`.
 */
export interface Party {
  kind: PartyKind;
  id: string;
}

/** An on-chain destination. The only external shape; an external SOURCE is unrepresentable. */
export interface ExternalDestination {
  network: string;
  /** Rendered in full wherever it is shown — see `FullAddress` (policy 13). */
  address: string;
}

/** Exactly one of the two: either an internal party or an external address. */
export interface PaymentDestination {
  internal?: Party;
  external?: ExternalDestination;
}

export interface OpenPaymentRequest {
  source: Party;
  destination: PaymentDestination;
  amount: Decimal;
  /** Required and bounded (≤ 500 bytes); it is what the approvers read. */
  reason: string;
}

// ── An end as the plane renders it back ────────────────────────────────────────

export type PaymentEndKind = PartyKind | "external";

/**
 * One end of an order, resolved for a reader.
 *
 * `label` is the canonical name the digest binds; `detail` is what a PERSON recognises the
 * end by — the receiving investor's masked mailbox, the product's title — so "investor
 * 8f3e…" is not approved for the wrong person. The BFF sends EMPTY STRINGS, not null, for
 * the fields a kind does not carry (`network`/`address` on an internal end, `detail` when
 * there is nothing to add).
 */
export interface PaymentEnd {
  label: string;
  kind: PaymentEndKind | string;
  id: string;
  network: string;
  address: string;
  detail: string;
}

// ── The two derived facts ──────────────────────────────────────────────────────

/** Where the money lands, by what it takes to get it there (docs/CONSILIUM.md § Three tiers). */
export type PaymentTier = "internal" | "service" | "external";

/** Who must agree, decided by the source alone. */
export type PaymentRequirement = "owner_consilium" | "subject_consent";

// ── The consent seat ───────────────────────────────────────────────────────────

/**
 * The one emailed seat a `subject_consent` order carries — the investor whose money it is.
 *
 * `decision` is the wire's spelling and goes through `settledConsent`
 * (`shared/lib/decision.ts`): an unanswered seat is the explicit `"pending"`, which is
 * truthy, and reading it as an answer is the bug that once hid every vote form.
 * `invalidated` is the pins moving (a token revoked, a mailbox changed) — the seat can no
 * longer be answered, whatever `decision` says.
 */
export interface PaymentConsent {
  subject_email: string;
  decision: string;
  notified: boolean;
  attempts_remaining: number;
  invalidated: boolean;
  invalidation_reason: string;
}

// ── The order ──────────────────────────────────────────────────────────────────

export interface Payment {
  id: string;
  /** pending · approved · executed · execution_failed · rejected · expired · cancelled — open. */
  state: string;
  tier: PaymentTier | string;
  source: PaymentEnd;
  destination: PaymentEnd;
  amount: Decimal;
  reason: string;
  requirement: PaymentRequirement | string;
  payload_hash: string;
  initiator_email: string;
  /** Set for an `owner_consilium` order — the room where the tally lives. */
  consilium_id: string | null;
  /** Set for a `subject_consent` order. */
  consent: PaymentConsent | null;
  created_at: Timestamp;
  expires_at: Timestamp;
  decided_at: Timestamp | null;
  /** The withdrawal an executed L1 order became; null for the internal tiers. */
  executed_withdrawal_id: string | null;
  /** Why an approved order did not move money. Terminal; retried by nothing. */
  failure_reason: string | null;
  version: number;
}

export interface PaymentList {
  items: Payment[];
}

// ── Public: the emailed consent ────────────────────────────────────────────────

export type ConsentDecision = "approve" | "reject";

/**
 * What `GET /api/approval/consent/:token` renders — the terms and nothing else. Same rules
 * as the payout approval: the GET is inert because mail scanners click every link, and
 * every dead token answers one identical 404 (policy 5, 10).
 */
export interface PaymentConsentInvitation {
  payment_id: string;
  state: string;
  tier: PaymentTier | string;
  source: PaymentEnd;
  destination: PaymentEnd;
  amount: Decimal;
  reason: string;
  payload_hash: string;
  initiator_email: string;
  /** The reader — whose money this is. */
  subject_email: string;
  expires_at: Timestamp;
  /** Wire spelling; read through `settledConsent`. */
  decision: string;
  attempts_remaining: number;
}

/** The answer to a cast consent. `invitation` is the authoritative re-read. */
export interface PaymentConsentResult {
  invitation: PaymentConsentInvitation;
  decided: boolean;
}

// ── The terms as a consilium carries them ──────────────────────────────────────

/**
 * What a payment consilium authorizes — the sibling of `RevenuePayout` on a `Consilium`
 * and on its emailed invitation. Two ends in words rather than a rail and an address; an
 * external destination still carries both inside `destination`.
 */
export interface ConsiliumPaymentTerms {
  payment_id: string;
  tier: PaymentTier | string;
  source: PaymentEnd;
  destination: PaymentEnd;
  amount: Decimal;
  reason: string;
}
